Packages/com.vrchat.UdonSharp/Editor/UdonSharpEditorUtility.cs

```diff
@@ -380,77 +379,261 @@ bool NeedsSerializationUpgrade(UdonBehaviour udonBehaviour)
             if (phase1FixupPrefabRoots.Count == 0 && phase2FixupPrefabRoots.Count == 0)
                 return;
             
-            if (phase2FixupPrefabRoots.Count > 0)
-                UdonSharpUtils.Log($"Running upgrade process on {phase2FixupPrefabRoots.Count} prefabs: {string.Join(", ", phase2FixupPrefabRoots.Select(e => e.name))}");
-            
-            using (new UdonSharpEditorManager.AssetEditScope())
+            UdonSharpPrefabDAG prefabDag = new UdonSharpPrefabDAG(prefabRoots);
+
+            // Walk up from children -> parents and mark prefab deltas on all U# behaviours to be upgraded
+            // Prevents versioning from being overwritten when a parent prefab is upgraded
+            foreach (string prefabPath in prefabDag.Reverse())
             {
-                foreach (GameObject prefabRoot in phase1FixupPrefabRoots)
+                GameObject prefabRoot = PrefabUtility.LoadPrefabContents(prefabPath);
+
+                bool needsSave = false;
+
+                try
                 {
-                    try
+                    HashSet<UdonBehaviour> behavioursToPrepare = new HashSet<UdonBehaviour>();
+                    
+                    foreach (UdonBehaviour behaviour in prefabRoot.GetComponentsInChildren<UdonBehaviour>(true))
                     {
-                        foreach (UdonBehaviour udonBehaviour in prefabRoot.GetComponentsInChildren<UdonBehaviour>(true))
+                        if (PrefabUtility.GetCorrespondingObjectFromSource(behaviour) != behaviour && (NeedsNewProxy(behaviour) || NeedsSerializationUpgrade(behaviour)))
                         {
-                            if (!NeedsNewProxy(udonBehaviour))
-                                continue;
-                            
-                            UdonSharpBehaviour newProxy = (UdonSharpBehaviour)udonBehaviour.gameObject.AddComponent(GetUdonSharpBehaviourType(udonBehaviour));
-                            newProxy.enabled = udonBehaviour.enabled;
+                            behavioursToPrepare.Add(behaviour);
+                        }
+                    }
+                    
+                    // Deltas are stored per-prefab-instance-root in a given prefab, don't question it. Thanks.
+                    // We take care to not accidentally hit any non-U#-behaviour deltas here
+                    // These APIs are not documented properly at all and the only mentions of them on forum posts are how they don't work with no solutions posted :))))
```

Comment:
Let me know if you'd like to submit a support request to get the official info from Unity on how to do any of this.

Reply:
I think it's figured out at this point. It would be good to ask Unity to document it properly though because there's a lot of things you can do with https://docs.unity3d.com/ScriptReference/PrefabUtility.SetPropertyModifications.html that 'work' most of the time but only by happenstance or don't properly cover all cases you'd see prefab modifications happen in. I can also submit an edit for the doc page on Unity but I'm not entirely sure that I'm doing it 100% as it should be done.